aten/src/ATen/native/mkldnn/Conv.cpp

```diff
@@ -35,43 +35,78 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
+Tensor _ideep_conv2d(
+    const Tensor& input, const Tensor& weight, const Tensor& bias,
+    at::IntArrayRef padding, at::IntArrayRef stride, at::IntArrayRef dilation, int64_t groups) {
 
-// Often written as 2 + max_dim (extra dims for batch size and channels)
-constexpr int max_dim = 3;
+  auto stride_vec = expand_param_if_needed(stride, 2);
+  auto padding_vec = expand_param_if_needed(padding, 2);
+  auto dilation_vec = expand_param_if_needed(dilation, 2);
 
-static std::vector<int64_t> conv_output_size(
-    IntArrayRef input_size, IntArrayRef weight_size,
-    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups)
-{
-  auto dim = input_size.size();
-  std::vector<int64_t> output_size(dim);
-  output_size[0] = input_size[input_batch_size_dim];
-  output_size[1] = weight_size[weight_output_channels_dim];
-  for (size_t d = 2; d < dim; ++d) {
-    auto kernel = dilation[d - 2] * (weight_size[d] - 1) + 1;
-    output_size[d] = (input_size[d] + (2 * padding[d - 2])
-                        - kernel) / stride[d - 2] + 1;
+  const ideep::tensor& x = itensor_from_mkldnn(input);
+  const ideep::tensor& w = itensor_from_mkldnn(weight);
+
+  const std::vector<int64_t> output_sizes = conv_output_size(
+    input.sizes(), weight.sizes(), padding, stride, dilation);
+  ideep::scale_t dummy_scale;
+  ideep::tensor y;
+  if (bias.defined()) {
+    const ideep::tensor& b = itensor_from_mkldnn(bias);
+    ideep::convolution_forward::compute<AllocForMKLDNN>(
+      x,
+      w,
+      b,
+      {output_sizes.cbegin(), output_sizes.cend()},
+      y,
+      {stride_vec.cbegin(), stride_vec.cend()},
+      {dilation_vec.cbegin(), dilation_vec.cend()},
+      {padding_vec.cbegin(), padding_vec.cend()},
+      {padding_vec.cbegin(), padding_vec.cend()},
+      groups,
+      dummy_scale,
+      dummy_scale,
+      dummy_scale,
+      ideep::descriptor_group::attr_t{},
+      ideep::algorithm::convolution_direct,
+      ideep::prop_kind::forward);
+  } else {
+    ideep::convolution_forward::compute<AllocForMKLDNN>(
+      x,
+      w,
+      {output_sizes.cbegin(), output_sizes.cend()},
+      y,
+      {stride_vec.cbegin(), stride_vec.cend()},
+      {dilation_vec.cbegin(), dilation_vec.cend()},
+      {padding_vec.cbegin(), padding_vec.cend()},
+      {padding_vec.cbegin(), padding_vec.cend()},
+      groups,
+      dummy_scale,
+      dummy_scale,
+      dummy_scale,
+      ideep::descriptor_group::attr_t{},
+      ideep::algorithm::convolution_direct,
+      ideep::prop_kind::forward);
   }
-  return output_size;
+  return new_with_itensor_mkldnn(std::move(y), input.options());
 }
 
 at::Tensor mkldnn_convolution(
     const at::Tensor& input, const at::Tensor& weight, const at::Tensor& bias,
     IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups)
 {
+  if (input.type_id() == MkldnnCPUTensorId()) {
+    return _ideep_conv2d(input, weight, bias, padding, stride, dilation, groups);
+  }
+
```

Comment:
1. the logic is for `MkldnnCPUTensorId()` go `_ideep_conv2d` and else (`CPUTensorId`) go `mkldnn_convolution`, actually it is not necessary and they can be merged into one function.
2. `ideep` is not a standalone library, it is a header that wraps up `mkldnn`. Their function are identical so there is no need to keep two sets of implementations. The current code may mislead people that `ideep` and `mkldnn` are two parallel libraries.
3. The existing code `mkldnn_convolution`, `mkldnn_convolution_backward_xxx` calls `mkldnn` directly. Since we need to switch them to `ideep`. I suggest switch them all and remove `Runtime.h` as well as other direct `mkldnn` callings. You may take #18460 as a reference.